lisa/_assets/kmodules/sched_tp/perf_counters.c

```diff
@@ -0,0 +1,694 @@
+
+	id = srcu_read_lock(&perf_data->srcu);
+	hlist_for_each_entry_srcu(entry, entry_list, node,
+				  srcu_read_lock_held(&perf_data->srcu)) {
+		value = perf_event_read_value(entry->event,
+					      &enabled,
+					      &running);
+		trace_perf_counter(cpu, entry->group->raw_id, value);
+	}
+	srcu_read_unlock(&perf_data->srcu, id);
+leave:
+	return perf_data->sampling_period;
+}
+
+static __maybe_unused
+void perfctr_reactivate_workers(struct perfctr_core *perf_data)
+{
+	int cpu;
+
+	if (!perf_data->nr_events)
+		return;
+
+	for_each_online_cpu(cpu) {
+		 struct work_item *work;
+
+		 work = per_cpu_ptr(perf_data->pcpu_data, cpu)->work;
+		 if (WARN_ON(!work)) /* hot-plug */
+		 	 continue;
+		 restart_work(work, perf_data->sampling_period);
+	}
+}
+
+static void perfctr_event_release_entry(struct perfctr_event_entry *entry);
+
+static int perfctr_event_activate_single(struct perfctr_core *perf_data,
+					 struct perf_event_attr *attr)
+{
+	struct perfctr_event_group *group;
+	struct perfctr_event_entry *entry;
+	struct hlist_node *next;
+	int cpu;
+
+	WARN_ON(!mutex_is_locked(&perf_data->lock));
+
+	group = kzalloc(sizeof(*group), GFP_KERNEL);
+	if (!group)
+		return -ENOMEM;
+
+	group->raw_id = PERF_COUNT_HW_MAX;
+
+	for_each_online_cpu(cpu) {
+		entry = kzalloc(sizeof(*entry), GFP_KERNEL);
+		if (!entry)
+			goto activate_failed;
+
+		entry->event =
+			/* No voerflow handler, at least not at this point */
+			perf_event_create_kernel_counter(attr, cpu, NULL,
+							 NULL, NULL);
+		if (IS_ERR(entry->event)) {
+			pr_err("Failed to create counter for %d on cpu%d\n",
+			       attr->config, cpu);
+			kfree(entry);
+			goto activate_failed;
+		}
+
+		hlist_add_head_rcu(&entry->node,
+				   &per_cpu_ptr(perf_data->pcpu_data, cpu)->events);
+		perf_event_enable(entry->event);
+
+		hlist_add_head(&entry->group_link, &group->entries);
+		entry->group = group;
+
+		/* One-time only */
+		if (group->raw_id != PERF_COUNT_HW_MAX)
+			continue;
+		if (attr->type == PERF_TYPE_RAW || !IS_ENABLED(CONFIG_ARM_PMU)) {
+			group->raw_id = attr->config;
+		} else {
+			struct arm_pmu *arm_pmu;
+
+			arm_pmu = to_arm_pmu(entry->event->pmu);
+			/* There needs to be a better way to do this !!*/
+			group->raw_id = arm_pmu->map_event(entry->event);
+		}
+	}
+	list_add_tail(&group->node, &perf_data->events);
+	++perf_data->nr_events;
+
+	return 0;
+
+activate_failed:
+	hlist_for_each_entry_safe(entry, next, &group->entries, group_link) {
+		hlist_del_rcu(&entry->node);
+		synchronize_srcu(&perf_data->srcu);
+		hlist_del(&entry->group_link);
+		perfctr_event_release_entry(entry);
+	}
+	kfree(group);
+	return -ENOMEM;
+
+}
+
+/* Lookup match type */
+enum perfctr_match_type {
+	PERFCTR_MATCH_NAME,
+	PERFCTR_MATCH_STATUS
+};
+
+struct perfctr_match {
+	void *data;
+	enum perfctr_match_type type;
+};
+
+static int perfctr_event_activate(struct perfctr_core *perf_data,
+				  const struct perfctr_match *match)
+{
+	int result = -EINVAL, i;
```

Comment:
I'd prefer splitting definitions each on their own line, especially if there is a mix of init and non-init